k2pow-service/src/main.rs

```diff
@@ -0,0 +1,377 @@
+
+    /// the number of cores to use. the optimal value depends
+    /// on the type of CPU used. `0` means use all cores.
+    #[arg(long, default_value = "0")]
+    cores: u8,
+
+    #[arg(long, default_value_t = RandomXMode::Fast)]
+    randomx_mode: RandomXMode,
+
+    /// allocate RandomX memory in large pages.
+    #[arg(long, default_value = "false")]
+    randomx_large_pages: bool,
+}
+
+/// RandomX modes of operation
+///
+/// They are interchangeable as they give the same results but have different
+/// purpose and memory requirements.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, ValueEnum, serde::Serialize)]
+enum RandomXMode {
+    /// Fast mode for proving. Requires 2080 MiB of memory.
+    Fast,
+    /// Light mode for verification. Requires only 256 MiB of memory, but runs significantly slower
+    Light,
+}
+
+impl std::fmt::Display for RandomXMode {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            RandomXMode::Fast => write!(f, "fast"),
+            RandomXMode::Light => write!(f, "light"),
+        }
+    }
+}
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let args = Cli::parse();
+
+    LogTracer::init()?;
+    let env_filter = EnvFilter::try_from_default_env().unwrap_or_else(|_| EnvFilter::new("INFO"));
+    let subscriber = FmtSubscriber::builder()
+        .with_env_filter(env_filter)
+        .finish();
+    tracing::subscriber::set_global_default(subscriber)?;
+
+    let job_manager = Arc::new(job_manager::JobManager::new());
+    _ = tokio::task::spawn(consume(
+        job_manager.clone(),
+        args.cores,
+        args.randomx_mode,
+        args.randomx_large_pages,
+    ));
+    let router = router(job_manager);
+    tracing::info!(
+        "starting http server with bind address: {}",
+        args.bind_address
+    );
+    let listener = tokio::net::TcpListener::bind(args.bind_address)
+        .await
+        .unwrap();
+    axum::serve(listener, router).await.unwrap();
+
+    Ok(())
+}
+
+fn router<T: GetOrCreate>(job_manager: Arc<T>) -> Router {
+    Router::new()
+        .route("/", get(root))
+        .route(
+            "/job/:miner/:nonce_group/:challenge/:difficulty",
+            get(get_job),
+        )
+        .with_state(job_manager)
+        .layer(
+            TraceLayer::new_for_http()
+                .make_span_with(|request: &Request<_>| {
+                    let matched_path = request.uri().to_string();
+
+                    info_span!(
+                        "http_request",
+                        method = ?request.method(),
+                        uri =  matched_path,
+                        status = tracing::field::Empty,
+                    )
+                })
+                .on_response(|response: &Response, _latency: Duration, span: &Span| {
+                    span.record("status", response.status().as_str());
+                    tracing::info!("served request");
```

Comment:
So I actually like this in `info` because you can see that the server is doing something. It tells you it serves requests and so you know it's still alive and generating output/serving requests. It doesn't show any debugging information, just that the request was served and the status code. I like to think of it as something similar to nginx access log

Reply:
Logs that tell that something happened and are emitted on every request are usually trace-level. They have no value for the operator except for debugging.